crates/bevy_ui/src/layout_components.rs

```diff
@@ -0,0 +1,430 @@
+/// Controls the size of UI nodes
+///
+/// Layout is performed according to the [`Layout`]
+/// To check the actual size of a UI element, read its [`Transform`](bevy_transform::components::Transform) component
+#[derive(Component, Copy, Clone, PartialEq, Debug, Default, Serialize, Deserialize, Reflect)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct SizeConstraints {
+    /// The minimum extent, which cannot be violated by the layouting algorithm
+    pub min: Size,
+    /// The suggested extent, which will be used if other constraints can be comfortably satisfied
+    pub suggested: Size,
+    /// The maximum extent, which cannot be violated by the layouting algorithm
+    pub max: Size,
+    /// The expected aspect ratio, computed as width / height
+    pub aspect_ratio: Option<f32>,
+}
+
+impl SizeConstraints {
+    pub const DEFAULT: SizeConstraints = SizeConstraints {
+        min: Size::DEFAULT,
+        suggested: Size::DEFAULT,
+        max: Size::DEFAULT,
+        aspect_ratio: None,
+    };
+
+    pub const FULL: SizeConstraints = SizeConstraints {
+        min: Size::DEFAULT,
+        suggested: Size::FULL,
+        max: Size::FULL,
+        aspect_ratio: None,
+    };
+
+    /// Sets only the minimum extent
+    pub const fn min(width: Val, height: Val) -> SizeConstraints {
+        SizeConstraints {
+            min: Size::new(width, height),
+            ..Self::DEFAULT
+        }
+    }
+
+    /// Sets only the suggested extent
+    pub const fn suggested(width: Val, height: Val) -> SizeConstraints {
+        SizeConstraints {
+            suggested: Size::new(width, height),
+            ..Self::DEFAULT
+        }
+    }
+
+    /// Sets only the suggested extent
+    pub const fn max(width: Val, height: Val) -> SizeConstraints {
+        SizeConstraints {
+            max: Size::new(width, height),
+            ..Self::DEFAULT
+        }
+    }
+}
+
+/// The space around and inside of a UI node
+#[derive(Component, Copy, Clone, PartialEq, Debug, Default, Serialize, Deserialize, Reflect)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct Spacing {
+    /// The space around the outside of the UI element
+    pub margin: UiRect,
+    /// The space around the inside of the UI element
+    pub padding: UiRect,
+    /// The space around the outside of the UI element that can be colored to create a visible border
+    pub border: UiRect,
+}
+
+impl Spacing {
+    pub const DEFAULT: Spacing = Spacing {
+        margin: UiRect::DEFAULT,
+        padding: UiRect::DEFAULT,
+        border: UiRect::DEFAULT,
+    };
+
+    /// Adds an automatic margin on all sides.
+    ///
+    /// When the element is explicitly sized, this will (hopefully) cause the element
+    /// to be vertically and horizontally centered relative to its parent.
+    pub const AUTO_MARGIN: Spacing = Spacing::margin(UiRect {
+        left: Val::Auto,
+        right: Val::Auto,
+        bottom: Val::Auto,
+        top: Val::Auto,
+    });
```

Comment:
```suggestion
    pub const AUTO_MARGIN: Spacing = Spacing::margin_all(Val::Auto);
```
I don't think this const reduces code duplication that much 🤷